File: src/core/data_handler.py
class DataHandler:
    CONTROL_CHAR_MAP = {
        0x00: '\u2400', 0x01: '\u2401', 0x02: '\u2402', 0x03: '\u2403',
        0x04: '\u2404', 0x05: '\u2405', 0x06: '\u2406', 0x07: '\u2407',
        0x08: '\u2408', 0x09: '\u2409', 0x0A: '\u240A', 0x0B: '\u240B',
        0x0C: '\u240C', 0x0D: '\u240D', 0x0E: '\u240E', 0x0F: '\u240F',
        0x10: '\u2410', 0x11: '\u2411', 0x12: '\u2412', 0x13: '\u2413',
        0x14: '\u2414', 0x15: '\u2415', 0x16: '\u2416', 0x17: '\u2417',
        0x18: '\u2418', 0x19: '\u2419', 0x1A: '\u241A', 0x1B: '\u241B',
        0x1C: '\u241C', 0x1D: '\u241D', 0x1E: '\u241E', 0x1F: '\u241F',
        0x7F: '\u2421',
    }

    @staticmethod
    def bytes_to_hex(data):
        return ' '.join(f'{b:02X}' for b in data)

    @staticmethod
    def bytes_to_ascii(data):
        result = []
        for b in data:
            if b in DataHandler.CONTROL_CHAR_MAP:
                result.append(DataHandler.CONTROL_CHAR_MAP[b])
                if b == 0x0A:
                    result.append('\n')
            elif 0x20 <= b < 0x7F:
                result.append(chr(b))
            else:
                result.append(f'\\x{b:02x}')
        return ''.join(result)

    @staticmethod
    def format_display(data, mode):
        if mode == 'HEX':
            return DataHandler.bytes_to_hex(data)
        elif mode == 'ASCII':
            return DataHandler.bytes_to_ascii(data)
        elif mode == 'MIXED':
            hex_str = DataHandler.bytes_to_hex(data)
            ascii_str = DataHandler.bytes_to_ascii(data)
            return f"\n \u2190 HEX: {hex_str}\n \u2190 ASCII: {ascii_str}"
        return ''

    @staticmethod
    def validate_hex_input(text):
        hex_str = text.replace(' ', '').replace('\n', '')
        try:
            bytes.fromhex(hex_str)
            return True
        except ValueError:
            return False

File: src/core/data_handler.cpp
#include "data_handler.h"

#include <QTextStream>

QMap<uchar, QChar> DataHandler::initControlCharMap() {
    QMap<uchar, QChar> map;
    // 0x00-0x1F -> U+2400-U+241F
    for (uchar i = 0; i < 0x20; ++i) {
        map.insert(i, QChar(0x2400 + i));
    }
    // 0x7F (DEL) -> U+2421
    map.insert(0x7F, QChar(0x2421));
    return map;
}

QChar DataHandler::controlCharMap(uchar ch) {
    static QMap<uchar, QChar> map = initControlCharMap();
    return map.value(ch, QChar(ch));
}

QString DataHandler::bytesToHex(const QByteArray &data) {
    if (data.isEmpty()) return QString();

    QString result;
    result.reserve(data.size() * 3);
    for (int i = 0; i < data.size(); ++i) {
        if (i > 0) result += ' ';
        result += QString::number(static_cast<uchar>(data[i]), 16).rightJustified(2, '0').toUpper();
    }
    return result;
}

QString DataHandler::bytesToAscii(const QByteArray &data) {
    if (data.isEmpty()) return QString();

    QString result;
    result.reserve(data.size());
    for (int i = 0; i < data.size(); ++i) {
        uchar ch = static_cast<uchar>(data[i]);
        if (ch == 0x0A) {
            // LF: append control symbol + real newline
            result += controlCharMap(ch);
            result += '\n';
        } else if (ch < 0x20 || ch == 0x7F) {
            // Control characters -> Unicode symbols
            result += controlCharMap(ch);
        } else if (ch >= 0x80) {
            // High bytes -> \xNN
            result += QStringLiteral("\\x%1").arg(ch, 2, 16, QChar('0')).toUpper();
        } else {
            // Printable ASCII
            result += QChar(ch);
        }
    }
    return result;
}

QString DataHandler::formatDisplay(const QByteArray &data, const QString &mode) {
    if (mode == "HEX") {
        return bytesToHex(data);
    } else if (mode == "ASCII") {
        return bytesToAscii(data);
    } else if (mode == "MIXED") {
        // Mixed: show both hex and ascii
        QString hex = bytesToHex(data);
        QString ascii = bytesToAscii(data);
        return hex + "  |  " + ascii;
    }
    return bytesToAscii(data);
}

QByteArray DataHandler::validateHexInput(const QString &text, bool *ok) {
    // Strip spaces and newlines
    QString cleaned = text;
    cleaned.remove(' ');
    cleaned.remove('\n');
    cleaned.remove('\r');
    cleaned.remove('\t');

    if (cleaned.isEmpty()) {
        if (ok) *ok = false;
        return QByteArray();
    }

    // Must have even number of hex chars
    if (cleaned.size() % 2 != 0) {
        if (ok) *ok = false;
        return QByteArray();
    }

    // Validate hex characters
    for (const QChar &ch : cleaned) {
        if (!ch.isLetterOrNumber()) {
            if (ok) *ok = false;
            return QByteArray();
        }
    }

    QByteArray result = QByteArray::fromHex(cleaned.toLatin1());
    if (result.isEmpty() && !cleaned.isEmpty()) {
        if (ok) *ok = false;
        return QByteArray();
    }

    if (ok) *ok = true;
    return result;
}
